src/k_smallest.rs

```diff
@@ -1,20 +1,230 @@
-use alloc::collections::BinaryHeap;
-use core::cmp::Ord;
+use alloc::vec::IntoIter;
+use core::cmp::{Ord, Ordering, Reverse};
+use core::mem::{replace, transmute, MaybeUninit};
+use core::ops::Range;
 
-pub(crate) fn k_smallest<T: Ord, I: Iterator<Item = T>>(mut iter: I, k: usize) -> BinaryHeap<T> {
-    if k == 0 { return BinaryHeap::new(); }
+fn k_smallest_dynamic<T, I: Iterator<Item = T>>(
+    iter: I,
+    k: usize,
+    order: impl Fn(&T, &T) -> Ordering,
+) -> IntoIter<T> {
+    let mut storage = Vec::new();
+    storage.resize_with(k, MaybeUninit::uninit);
+    let Range { end, .. } = capped_heapsort(iter, &mut storage, order);
+    storage.truncate(end);
+    let initialized: Vec<_> = unsafe { transmute(storage) };
```

Comment:
Sidenote: Is this `transmute` even allowed? (I think it's not.) [The docs](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#layout-1) state that the layout of `MaybeUninit<T>` is the same as the one for `T`, but that types containing `MaybeUninit<T>`/`T` do *not* necessarily have the same layout. (Rust apparently could e.g. reorder members of `Vec<MaybeUninit<T>>` in another way than the members of `Vec<T>`.) It seems the way out would be [`from_raw_parts`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.from_raw_parts), but I'd rather avoid `unsafe` in the first place.

Reply:
I suspect you're right that the transmute isn't valid and `from_raw_parts` would be needed, but as mentioned below, the new implementation is entirely safe.